R doc 1359 graph (#1041)
ravendb/docs: Documentation/4.2/Raven.Documentation.Pages/indexes/querying/graph/graph-queries-overview.markdown

﻿# Graph Querying: Overview  

---

{NOTE: }

RavenDB's [experimental](../../../indexes/querying/graph/graph-queries-overview#enabling-graph-querying) **graph** support 
allows you to query your database as if it had a graphical structure, gaining extreme efficiency and speed in recognizing 
relations between data elements and organizing them into searchable patterns. Intricate relationships that would render 

> a relational databases useless, become the asset they are meant to be.  

Reviewer: nice

* **No need for preliminary preparations**  
  You do not need to alter your database's structure or contents in order to start using graph queries. 
  Your existing collections and documents are used as graph elements, and their relations are inferred from documents' contents.  

* **Simple and effective syntax**  

>   We've integrated graph support into [RQL](), to make its learning and usage accessible and intuitive for any user, 

Reviewer: missing RQL link

Reviewer 2: fixed

  especially those already familiar with our query language.  

* **Comprehensive support**  
  Queries can be constructed by either clients using API methods or manually using the Studio, and executed 

>   by your distributed server. Results can be used by your clients, or shown textually and graphically by the Studio.  

Reviewer: `distributed server` is an odd way of phrasing this

Reviewer 2: changed to "distributed database"

* **..and large data volumes**   
  [Large, complex data volumes](https://www.datamation.com/big-data/big-data-companies.html) represent 
  an important step in the evolution of data management, and are evidently here to stay and develop.  
  As relational databases and the data model they enable are inefficient in (and often incapable of) 
  searching and managing big-data volumes with intricate relations, various applications take part 
  in complementing or replacing them.  
  Databases capable of running graph queries are a major contribution in this regard, though 
  not limited to the  management of large data volumes and often as comfortable and efficient 
  in handling smaller ones.  
* **..and Multi Model**  
  It is common to find graph querying as one of the features of a multi-model database, based upon 
  or cooperating with other database features.  
  RavenDB's graph capabilities are founded upon a capable document store, and data already deposited in 
  the store can participate graph querying with no preceding arrangements, easing user administration and 
  improving internal logic and data management.  

{NOTE: }

> ###Enabling Graph Querying

Reviewer 3: I think this is important enough to have its own glossary page (for enabling experimental features) and just point to it.

Reviewer: Yes, agreed. Also, the URL below is 4.1.1, and it really bugged me

Reviewer 2: done

Graph Querying is an **Experimental feature**, still under development. We are happy to provide it, and would be grateful 
to hear from you regarding your experiences with it.  
As other experimental fatures, it is disabled by default. You can enable it following theis simple procedure:  

* Open the RavenDB server folder, e.g. C:\Users\Dave\Downloads\RavenDB-4.1.1-windows-x64\Server  
* Open settings.json for editing  
* Enable the Experimental Features.  
   * Verify that the json file contains the following line:  
     "Features.Availability": "Experimental"  
   * Save settings.json and restart RavenDB Server.  

{NOTE/}

{PANEL/}

> {PANEL: Designing Graph Queries}  

Reviewer: I would actually open up here with some explanation of how to do graph queries and exploring in the studio, so they can understand

Reviewer 2: very true.
done.

####Graph representations

Graph querying enhances RQL with simple vocabulary and syntax that allow you to approach your existing data 
as if it had been designed graphically. Here's a basic query that shows relations between employees, using documents 
taken from the Northwind database RavenDB lets you install as sample data.  

* Graph query:  
{CODE-BLOCK:JSON}
match(employees as employee)-[ReportsTo as reportsTo]->(employees as incharge)  
{CODE-BLOCK/}
* Query results are provided by the Studio both graphically and textually.  

 ![Illustrative graph reqults](images/Overview_GraphicalView_1.png)  
 ![Textual graph results](images/Overview_TextualView_1.png)  

---

####Basic Terms, Syntax and Vocabulary

* **Graph Elements**  

>   Data elements ("Nodes") and their relations ("Edges") are represented in a graph as equally important.  

Reviewer: ```suggestion
  Documents ("Nodes") and their relations ("Edges") are represented in a graph as equally important.  
```

Reviewer 2: done

>   * **Data Nodes\***  

Reviewer: ```suggestion
  * **Nodes\***  
```

Reviewer 2: there's a comment there saying "We use the term "data nodes" to distinguish them from cluster servers ("cluster nodes")"

>     A "data node" can be a documents collection, or a subset of selected documents.  

Reviewer 3: A node could also be a result of an index query, which typically is a set of documents, but could also be the result of a map-reduce index

Reviewer: ```suggestion
    A "node" can be a documents collection, or a subset of selected documents.  
```

Reviewer: A node is a _result_ of a query, not a collection. That would be _nodes_.

Reviewer 2: done

>     **\*** _We use the term "data nodes" to make it easier for you to distinguish between the data elements 

Reviewer: Use vertex / edge terminology, then

    we talk about here, and servers of a cluster (that are also called "nodes")._  
  * **Edges**  
    An "edge" is a link between nodes, that joins them in a relation of some sort.  

>     A RavenDB edge is simply a string-field within a document, that refers to the unique identifier of a document.  

Reviewer 3: an edge could be a nested JSON for exmple we consider an order line to be an edge while the relationship is the product property the entire order line is the edge and you could query it,

Reviewer 2: it's explained when we talk about paths. I'll update the vocabulary.

Reviewer 2: done

    RavenDB's edges are always **directional**, pointing from one data node to another.  
     
* **Graph results**  

>   Here are the results of a very simple query.  

Reviewer: Show the query

  ![Figure 1. Simple Relation](images/Overview_Elements.png)  

  1. The first data node is **Dogs/Ruffus**, a document named **Ruffus** in the **Dogs** collection.  
    This is how the document may look like:
    {CODE-BLOCK:JSON}
    Document: Dogs/Ruffus
    {
       "Owner": "Owners/John",
       "Name": "...",

>        "@metadata": {

Reviewer: fomatting

       "@collection": "Dogs",
       "@id": "Dogs/Ruffus"}
    }
    {CODE-BLOCK/}

  2. The arrow titled **ownedBy** is the edge, indicating that Ruffus belongs to John.  
     You can find its definition in the **Ruffus** document as the "Owner" field containing John's ID.  

  3. The second data node is **Owners/John**, a document named **John** in the **Owners** collection.
    {CODE-BLOCK:JSON}
    Document: Owners/John
    {
       "Name": "...",
       "@metadata": {

>        "@collection": "Owners",

Reviewer: formatting

     Giving elements aliases isn't obligatory when they are defined [implicitly](../../../indexes/querying/graph/graph-queries-basic#implicitly-defining-nodes-and-edges), and **is** required when defining them [explicitly](../../../indexes/querying/graph/graph-queries-basic#explicitly-defining-data-elements).  
     It is, however, often recommended and sometimes essential.  
      * In our sample query, the Owner relation between a dog and its owner is given the alias "ownedBy", 
        because we wanted to emphasize this aspect of the relations. Another query may emphasize a different 
        aspect by using the alias "occupant", "patient" or something else.  
      * The same node or edge may appear multiple times in a query, sometimes in very different roles.  
        Using different aliases may be technically needed in such cases.  
      * To eliminate an entity from the results, use a sequence of `_` symbols as an alias (i.e. `_`, `__`, `___`..)  
      * Each alias needs to be unique.  
        Note that this is true for `_` aliases as well: use each `_` sequence (`_`, `__`, `___`..) only once.  
* **Graph Queries Flow**
   * Lucene indexing  

>      When a graph query is executed, the first thing RavenDB does is index each node clause using Lucene.  

Reviewer 3: not all nodes come out of Lucene, we have collection queries that don't go through Lucene, maybe a more accurate thing to say is that the graph query first query the node clause where query could result in either a Lucene query or a collection query.

Reviewer 2: "When a graph query is executed, RavenDB first queries data nodes.  
Resulting Lucene and Collection queries are indexed, and the indexes are grouped in tables the graph engine may use. "

     The result is a group of indexed tables that the graph engine can easily play with.  
   * Handling relations  
     If the query comprises edges, the graph engine uses them now while going through the table prepared during the first phase 
     and fathoming the relations between table elements.  

>      [Be aware](../../../indexes/querying/graph/graph-queries-basic#graph-queries-and-indexes) that this part of a query is performed in memory and is not indexed, so reruns actually mean 

Reviewer: That wouldn't actually always be the case. And this sounds really scary.
I'm not sure how this is different from any other query that may do some work

Reviewer 2: Removed "Be aware that this part of a query is performed in memory and is not indexed, so reruns actually mean re-running it".

     re-running it.  

{PANEL/}

{PANEL: FAQ}

####Q: When should or shouldn't I use graph queries?  

A: There are configurations and situations for which graph querying is an optimal solution, and other 
circumstances that require different approaches. You may find this list helpful in determining whether
to give it a go.  

* **Use graph querying when -**  
   * **Relations between data elements are a concern**.  

>      If your data continuously grows in quantity and intricity, queries become increasingly complicated and results 

Reviewer 3: I don't know if we want to encourage graph queries for a large data set, users should start with nodes that are already well-defined queries with a retelative small result set.

Reviewer: That is discussed below, but I agree. We want to traverse large amount of data, but not start from it, I think

Reviewer 2: "When data interrelations become intricate enough to complicate and slow down your queries, graph queries may well be a good solution."
